feat: linkdex report api

olizilla opened the pull request:
adds just enough lambda and infra to roll out a linkdex report to give us dag completeness over one or more cars for a user upload.

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: linkdex report api

adds just enough lambda and infra to roll out a linkdex report to give us dag completeness over one or more cars for a user upload.

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: add package lock

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: we only need one s3

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

alanshaw commented on README.md:
```diff
@@ -0,0 +1,39 @@
+
+Give it an s3 key to a CAR you just upladed as `key`, and it tells you if we have a complete DAG for the root CID.
+
+We check the key provided exists first. If you know the path and hash of CAR then you are probably dotStoarage.
```
```suggestion
We check the key provided exists first. If you know the path and hash of CAR then you are probably dotStorage.
```

vasco-santos replied:
```suggestion
We check if the key provided exists first. If you know the path and hash of CAR then you are probably dotStorage.
```

alanshaw commented on services/functions/linkdex.ts:
```diff
@@ -0,0 +1,81 @@
+  return {
+    ...opts,
+    statusCode: opts.statusCode ?? 200,
+    headers: Object.assign({ "Content-Type": "application/json" }, opts.headers),
```
```suggestion
    headers: Object.assign({ 'Content-Type': 'application/json' }, opts.headers),
```

alanshaw commented on services/functions/linkdex.ts:
```diff
@@ -0,0 +1,81 @@
+    const batch = data.Contents ?? []
+    for (const item of batch) {
+      if (item.Key !== undefined) {
+        files.push(item.Key)
```
Ensure suffix is `.car` to avoid adding `.car.idx` files?

olizilla replied:
YES! Very wise.

alanshaw commented on services/functions/linkdex.ts:
```diff
@@ -0,0 +1,81 @@
+  try {
+    const res = await s3.send(new HeadObjectCommand({ Key: key, Bucket: bucket }))
+    return true
+  } catch (err) {
+    false
```
```suggestion
  } catch {
    return false
```

alanshaw commented on services/functions/linkdex.ts:
```diff
@@ -0,0 +1,81 @@
+import { LinkIndexer } from 'linkdex'
+
+export const handler: APIGatewayProxyHandlerV2 = async (event) => {
+  const key = event.queryStringParameters?.key ?? ''
```
Accept multiple `key`? In uploadsV2 CAR parts are unlikely to be in the same directory and so we should be able to ask if a DAG is complete in the _given_ CARs.

olizilla replied:
Yes! But not just yet. I think this will change pretty significantly to be useful for uploads v2.

alanshaw commented on services/functions/linkdex.ts:
```diff
@@ -0,0 +1,81 @@
+  if (!key) {
+    return response({ error: { details: 'key query param is required' }}, { statusCode: 400 })
+  }
+  if (!key.startsWith('raw/') || !key.endsWith('.car') || key.split('/').length !== 4) {
```
Why restrict to `raw/` prefix and 4 dirs deep? AFAIK uploadsV2 will be `car-cid/car-cid.car`.

olizilla replied:
This initial implementation is focused on the existing upload flow. We can generalise it once it has proved to be useful. We will likely change this thing quite a bit to be useful to uploads v2

alanshaw commented on services/functions/linkdex.ts:
```diff
@@ -0,0 +1,81 @@
+  }
+  const carKeys = await listSiblings(key, bucket, s3)
+  const index = new LinkIndexer()
+  for (const carKey of carKeys) {
```
Add concurrency to reduce runtime and cost...?

olizilla replied:
yes like dis? https://github.com/web3-storage/linkdex-api/pull/1/commits/32baa8d2871caf55ac3415bcf7153b68468be390

alanshaw reviewed (approved):
This LGTM.

I was thinking about this slightly differently:

1. Lambda that triggers on object create in bucket
2. For each block insert `cid => [linkCid0, linkCid1]` into DB where `linkCid0` is non-raw.
3. API is `https://linkdex.api?cid=bafy...`
4.
    ```js
    let cids = [event.queryStringParameters?.cid]
    while true {
      try {
        const cid = cids.pop()
        if (!cid) return 'Complete'
        const links = db.select(cid)
        cids.push(...links)
      } catch (err) {
        if (err.is404) {
          return 'Partial'
        }
        throw err
      }
    }
    ```

vasco-santos reviewed (commented):
This is getting super cool

olizilla committed:
Apply suggestions from code review

Co-authored-by: Vasco Santos <santos.vasco10@gmail.com>
Co-authored-by: Alan Shaw <alan.shaw@protocol.ai>

olizilla committed:
fix: ignore non-car files

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: read CARs in parallel and retry car fetching

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: pass bucketName as env

also fix lambda config
- no timeout. run till you can run no more
- use arm for cheaper

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>
